src/main/java/com/vunet/keycloak/auth/otp/OtpAuthenticatorForm.java

```diff
@@ -0,0 +1,281 @@
+
+            Properties prop = new Properties();
+            prop.load(fis);
+            logger.info(prop.getProperty("Url from property file "+ URL_PROP));
+
+	        String username = inputData.getFirst(AuthenticationManager.FORM_USERNAME);
+	        String otp = inputData.getFirst(OTP);
+	
+	        OkHttpClient httpClient = new OkHttpClient();
+	
+	    	RequestBody formBody = new FormBody.Builder()
+	    		      .add(UNAME, username)
+	    		      .add(UOTP, otp)
+	    		      .add(TOKEN, prop.getProperty(TOKEN_PROP))
+	    		      .build();
+	
+		    Request request = new Request.Builder()
+		      .url(prop.getProperty(URL_PROP))
+		      .addHeader("Content-Type", "application/json")
+		      .addHeader("Cache-Control", "no-cache")
+		      .post(formBody)
+		      .build();
+
+
+			okhttp3.Response response = httpClient.newCall(request).execute();
+			JSONObject jsonObject = new JSONObject(response.toString());
+			int responseCode = jsonObject.getInt(RESPONSE_CODE);
+			
+			if(responseCode==200) {
+				boolean status = jsonObject.getBoolean(STATUS);
+				boolean userRegistered = jsonObject.getBoolean(USER_REGISTERED);
+				if(status&&userRegistered) {
+					return true;
+				}else if(!status&&userRegistered) {
+					context.getEvent().error(Errors.INVALID_CODE);
+		            Response challengeResponse = challenge(context, Messages.INVALID_TOTP);
+		            context.failureChallenge(AuthenticationFlowError.INVALID_CREDENTIALS, challengeResponse);
+		            return false;
+				}else if(!status&&!userRegistered) {
+					context.getEvent().error(Errors.USER_NOT_FOUND);
+		            Response challengeResponse = challenge(context, "User not registered");
+		            context.failureChallenge(AuthenticationFlowError.INVALID_USER, challengeResponse);
+		            return false;
+				}
+			} else if(responseCode==401){
+				String status = jsonObject.getString(STATUS);
+				String userRegistered = jsonObject.getString(USER_REGISTERED);
+				if(status.equalsIgnoreCase("InvalidToken")) {
+					context.getEvent().error(Errors.INVALID_TOKEN);
+		            Response challengeResponse = challenge(context,Messages.INTERNAL_SERVER_ERROR);
+		            context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR, challengeResponse);
+		            return false;
+				}
+			}
+			
+		} catch (IOException e) {
+			logger.error("Exception in validating OTP ",e);
+		}
+	    Response challengeResponse = challenge(context,Messages.INTERNAL_SERVER_ERROR);
+        context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR, challengeResponse);
+		return false;
+	}
+    
+    @Override
+    public boolean requiresUser() {
+        return false;
+    }
+
+    @Override
+    public boolean configuredFor(KeycloakSession session, RealmModel realm, UserModel user) {
+        return true;
+    }
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
```

Comment:
What are these NOOP functions? Why do we need to define them?

Reply:
We need to override certain methods by default when implement an interface. If not it will throw compilation error. We need to define our logic there. But in our case there is no logic needed. So it is just given as empty function

Reply:
Understood. Thanks